lib/services/role_test.go

```diff
@@ -3302,8 +3304,163 @@ func TestApplyTraits(t *testing.T) {
+						Namespace: "{{external.namespace}}",
+						Kind:      "*",
+						Name:      "name-*",
+						Verbs:     []string{"get", "list"},
+					},
+				},
+				outKubeResources: []types.KubernetesResource{
+					{
+						Namespace: "default",
+						Kind:      "*",
+						Name:      "name-*",
+						Verbs:     []string{"get", "list"},
+					},
+					{
+						Namespace: "kubens1",
+						Kind:      "*",
+						Name:      "name-*",
+						Verbs:     []string{"get", "list"},
+					},
+					{
+						Namespace: "kubens2",
+						Kind:      "*",
+						Name:      "name-*",
+						Verbs:     []string{"get", "list"},
+					},
+				},
+			},
+		},
+		{
+			comment: "full kubernetes resources replace rules",
+			inTraits: map[string][]string{
+				"namespace": {"kubens1", "kubens2"},
+				"names":     {"kubepod1", "kubepod2"},
+				"verbs":     {"update", "delete"},
+			},
+			allow: rule{
+				inKubeResources: []types.KubernetesResource{
+					{
+						Namespace: "{{external.namespace}}",
+						Kind:      "pod",
+						Name:      "{{external.names}}",
+						Verbs:     []string{"{{external.verbs}}", "list"},
+					},
+				},
+				outKubeResources: []types.KubernetesResource{
+					{
+						Namespace: "kubens1",
+						Kind:      "pod",
+						Name:      "kubepod1",
+						Verbs:     []string{"update", "delete", "list"},
+					},
 
+					{
+						Namespace: "kubens1",
+						Kind:      "pod",
+						Name:      "kubepod2",
+						Verbs:     []string{"update", "delete", "list"},
+					},
+
+					{
+						Namespace: "kubens2",
+						Kind:      "pod",
+						Name:      "kubepod1",
+						Verbs:     []string{"update", "delete", "list"},
+					},
+					{
+						Namespace: "kubens2",
+						Kind:      "pod",
+						Name:      "kubepod2",
+						Verbs:     []string{"update", "delete", "list"},
+					},
+				},
+			},
+		},
+		{
+			comment: "full deny kubernetes resources replace rules",
+			inTraits: map[string][]string{
+				"namespace": {"kubens1", "kubens2"},
+				"names":     {"kubepod1", "kubepod2"},
+				"verbs":     {"update", "delete"},
+			},
+			deny: rule{
+				inKubeResources: []types.KubernetesResource{
+					{
+						Namespace: "{{external.namespace}}",
+						Kind:      "pod",
+						Name:      "{{external.names}}",
+						Verbs:     []string{"{{external.verbs}}", "list"},
+					},
+				},
+				outKubeResources: []types.KubernetesResource{
+					{
+						Namespace: "kubens1",
+						Kind:      "pod",
+						Name:      "kubepod1",
+						Verbs:     []string{"update", "delete", "list"},
+					},
+					{
+						Namespace: "kubens1",
+						Kind:      "pod",
+						Name:      "kubepod2",
+						Verbs:     []string{"update", "delete", "list"},
+					},
+					{
+						Namespace: "kubens2",
+						Kind:      "pod",
+						Name:      "kubepod1",
+						Verbs:     []string{"update", "delete", "list"},
+					},
+					{
+						Namespace: "kubens2",
+						Kind:      "pod",
+						Name:      "kubepod2",
+						Verbs:     []string{"update", "delete", "list"},
```

Comment:
nit:  add a const for these verbs and reuse here and above?

`const verbUpdateListDelete = []string{"update", "delete", "list"}`
```suggestion
						Verbs:     verbUpdateListDelete,
```